File: doodlestudio/engine/timeline.py
"""Timeline layout shared by real audio assembly and synthetic previews.

Given each beat's measured clip length and per-character speech times, place
beats on the master clock, insert chapter gaps, take reading holds and gem
transition time, then derive captions, chapters, music intervals and the end card.
"""
from __future__ import annotations

import re

from . import captions as cap
from .storyboard import normalize

FPS = 30
CHAPTER_GAP = .6
TRANSITION = 2.8          # pull back 0.35 + fly/pin 0.9 + check 0.7 + circle 0.8 (+ slack)
END_CARD = 5.0            # pan to the closing page, write it, and let it be read
ZH_DWELL = .5             # extra reading pause per Mandarin paragraph (9/19 precedent)
ZOOM_IN = .9              # first part of each section: zoom into its agenda card
AGENDA_CARD = 2.6         # hand time per agenda card: the agenda holds until every card is drawn


def take_hold(beat, lang):
    head = (beat.get('take') or {}).get('headline', {}).get(lang, '')
    if lang == 'en':
        return max(3.0, len(head.split()) / 3.0)
    return max(3.0, len(re.findall(r'[一-鿿A-Za-z0-9]', head)) / 6.0)


def layout(episode, lang, clips):
    """clips[beat_id] = {'speech': seconds of speech incl. trailing clip gap, 'char_times': [...]}"""
    episode = normalize(episode)
    beats = episode['beats']
    chapters = {c['id']: c for c in episode['chapters']}
    cursor = 0.
    out_beats, order, transitions, capts = {}, [], [], []
    n_cards = sum(c['kind'] == 'section' for c in episode['chapters'])
    agenda_start = None
    for i, beat in enumerate(beats):
        clip = clips[beat['id']]
        start = cursor
        speech_end = start + clip['speech']
        end = speech_end + (ZH_DWELL if lang == 'zh' else 0.)
        nxt = beats[i + 1] if i + 1 < len(beats) else None
        chapter_change = nxt is not None and nxt['chapter'] != beat['chapter']
        info = {'start': round(start, 4), 'speech_end': round(speech_end, 4), 'char_times': clip['char_times']}
        if beat.get('kind') == 'take' and chapters[beat['chapter']]['kind'] == 'section':
            hold = take_hold(beat, lang)
            hold_end = end + hold
            t_end = hold_end + TRANSITION
            transitions.append({'section': beat['chapter'], 'take_beat': beat['id'], 'speech_end': round(end, 4),
                                'hold_end': round(hold_end, 4), 'end': round(t_end, 4),
                                'next': nxt['chapter'] if nxt else None})
            end = t_end
        elif chapter_change:
            end += CHAPTER_GAP
        if chapters[beat['chapter']]['kind'] == 'agenda':
            agenda_start = start if agenda_start is None else agenda_start
            if chapter_change:                           # short agenda narration: wait for the cards
                end = max(end, agenda_start + 1.4 + AGENDA_CARD * n_cards + CHAPTER_GAP)
        info['end'] = round(end, 4)
        out_beats[beat['id']] = info
        order.append(beat['id'])
        ct = clip['char_times']

        def char_time(pos, ct=ct):
            return ct[min(max(pos, 0), len(ct) - 1)] if ct else 0.
        for a, b, text in cap.cues_for_beat(beat['spoken'][lang], beat['display'][lang], lang, char_time,
                                           clip['speech'] - .15):
            capts.append({'start': round(start + a, 4), 'end': round(start + b, 4), 'text': text})
        cursor = end
    duration = cursor + END_CARD
    duration = round(-(-duration * FPS // 1) / FPS, 6)
    chaps = []
    for c in episode['chapters']:
        ids = [b['id'] for b in beats if b['chapter'] == c['id']]
        label = (c.get('label') or {}).get(lang, '').strip()
        title = (c.get('title') or {}).get(lang, '').strip()
        chaps.append({'id': c['id'], 'start': out_beats[ids[0]]['start'], 'end': out_beats[ids[-1]]['end'],
                      'title': f'{label} · {title}' if c['kind'] not in ('intro', 'outro', 'agenda') and label
                      and title and title.lower() != label.lower() else (label or title)})
    # Music: flagged beats, section transitions (pull-back onward) and the end card.
    music = []
    for beat in beats:
        if beat.get('music'):
            info = out_beats[beat['id']]
            music.append([info['start'], info['end']])
    for tr in transitions:
        music.append([tr['hold_end'], tr['end']])
    music.append([duration - END_CARD, duration])
    music.sort()
    merged = []
    for a, b in music:
        if merged and a <= merged[-1][1] + 1.0:
            merged[-1][1] = max(merged[-1][1], b)
        else:
            merged.append([a, b])
    return {'language': lang, 'fps': FPS, 'duration': duration, 'beats': out_beats, 'beat_order': order,
            'captions': capts, 'chapters': chaps, 'transitions': transitions,
            'music': [{'start': round(a, 4), 'end': round(b, 4)} for a, b in merged],
            'end_card': {'start': round(duration - END_CARD, 4), 'end': duration}}


def synthetic_clips(episode, lang):
    """Reading-rate estimate for previews only (EN ~2.45 words/s, ZH ~4.4 chars/s)."""
    clips = {}
    for beat in episode['beats']:
        text = beat['spoken'][lang]
        if lang == 'en':
            seconds = max(2.0, len(text.split()) / 2.45)
        else:
            seconds = max(2.0, len(re.findall(r'[一-鿿]', text)) / 4.4 + len(re.findall(r'[A-Za-z]+', text)) * .25)
        n = len(text)
        clips[beat['id']] = {'speech': seconds + .45, 'char_times': [round(seconds * k / max(1, n), 3) for k in range(n)]}
    return clips

File: edu/extension/engine/timeline.js
// Timeline layout (port of engine/timeline.py). Given each beat's measured clip length and per-character speech
// times, place beats on the master clock, insert chapter gaps, take reading holds and transition time, then derive
// captions, chapters, music intervals and the end card. `pauses` (beat id -> seconds, from pacing) add silence after a
// beat so the narration waits for the drawing hand; a takeaway beat starts TAKE_PREROLL after the beat before it.
import { cuesForBeat } from './captions.js';
import { normalizeStoryboard } from './storyboard.js';

export const FPS = 30;
export const CHAPTER_GAP = 0.6;
export const TRANSITION = 2.8;
export const END_CARD = 5.0;
export const ZOOM_IN = 0.9;
export const AGENDA_CARD = 2.6;
export const TAKE_PREROLL = 2.0;
const r4 = (x) => Math.round(x * 1e4) / 1e4;

export function takeHold(beat) {
  const head = beat.take?.headline?.en || '';
  return Math.max(3.0, head.split(/\s+/).filter(Boolean).length / 3.0);
}

/** clips[id] = { speech: seconds incl. the gap after it, charTimes: [...] }; pauses[id] = silence after that beat. */
export function layout(episode, clips, pauses = {}, measure = null) {
  episode = normalizeStoryboard(episode);
  const beats = episode.beats;
  const chapters = Object.fromEntries(episode.chapters.map((c) => [c.id, c]));
  let cursor = 0;
  const outBeats = {};
  const order = [];
  const transitions = [];
  const capts = [];
  const nCards = episode.chapters.filter((c) => c.kind === 'section').length;
  let agendaStart = null;
  beats.forEach((beat, i) => {
    const clip = clips[beat.id];
    const take = beat.kind === 'take' && chapters[beat.chapter].kind === 'section';
    const prep = cursor;
    const start = cursor + (take ? TAKE_PREROLL : 0);
    const speechEnd = start + clip.speech;
    let end = speechEnd + Number(pauses[beat.id] || 0);
    const nxt = beats[i + 1] || null;
    const chapterChange = nxt !== null && nxt.chapter !== beat.chapter;
    const info = { start: r4(start), speech_end: r4(speechEnd), char_times: clip.charTimes };
    if (take) {
      info.prep = r4(prep);
      const holdEnd = end + takeHold(beat);
      const tEnd = holdEnd + TRANSITION;
      transitions.push({ section: beat.chapter, take_beat: beat.id, speech_end: r4(end), hold_end: r4(holdEnd), end: r4(tEnd),
        next: nxt ? nxt.chapter : null });
      end = tEnd;
    } else if (chapterChange) {
      end += CHAPTER_GAP;
    }
    if (chapters[beat.chapter].kind === 'agenda') {
      agendaStart = agendaStart === null ? start : agendaStart;
      if (chapterChange) end = Math.max(end, agendaStart + 1.4 + AGENDA_CARD * nCards + CHAPTER_GAP);   // wait for the cards
    }
    info.end = r4(end);
    outBeats[beat.id] = info;
    order.push(beat.id);
    if (measure) {
      const ct = clip.charTimes;
      const charTime = (pos) => (ct.length ? ct[Math.min(Math.max(pos, 0), ct.length - 1)] : 0);
      for (const [a, b, text] of cuesForBeat(beat.spoken.en, beat.display.en, charTime, clip.speech - 0.15, measure)) {
        capts.push({ start: r4(start + a), end: r4(start + b), text });
      }
    }
    cursor = end;
  });
  let duration = cursor + END_CARD;
  duration = Math.round((Math.ceil(duration * FPS) / FPS) * 1e6) / 1e6;
  const chaps = episode.chapters.map((c) => {
    const ids = beats.filter((b) => b.chapter === c.id).map((b) => b.id);
    const label = (c.label?.en || '').trim();
    const title = (c.title?.en || '').trim();
    return { id: c.id, start: outBeats[ids[0]].start, end: outBeats[ids[ids.length - 1]].end,
      title: !['intro', 'outro', 'agenda'].includes(c.kind) && label && title && title.toLowerCase() !== label.toLowerCase()
        ? `${label} · ${title}` : (label || title) };
  });
  const music = [];
  for (const beat of beats) if (beat.music) music.push([outBeats[beat.id].start, outBeats[beat.id].end]);
  for (const tr of transitions) music.push([tr.hold_end, tr.end]);
  music.push([duration - END_CARD, duration]);
  music.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const merged = [];
  for (const [a, b] of music) {
    if (merged.length && a <= merged[merged.length - 1][1] + 1.0) merged[merged.length - 1][1] = Math.max(merged[merged.length - 1][1], b);
    else merged.push([a, b]);
  }
  return { language: 'en', fps: FPS, duration, beats: outBeats, beat_order: order, captions: capts, chapters: chaps,
    transitions, pauses: Object.fromEntries(Object.entries(pauses).filter(([, v]) => v).map(([k, v]) => [k, Math.round(Number(v) * 1000) / 1000])),
    music: merged.map(([a, b]) => ({ start: r4(a), end: r4(b) })), end_card: { start: r4(duration - END_CARD), end: duration } };
}

/** Reading-rate estimate for previews and tests (~2.45 words/s). */
export function syntheticClips(episode) {
  const clips = {};
  for (const beat of episode.beats) {
    const text = beat.spoken.en;
    const seconds = Math.max(2.0, text.split(/\s+/).filter(Boolean).length / 2.45);
    const n = text.length;
    clips[beat.id] = { speech: seconds + 0.45, charTimes: Array.from({ length: n }, (_, k) => Math.round(seconds * k / Math.max(1, n) * 1000) / 1000) };
  }
  return clips;
}
